### PR TITLE
SQL: use the correct data type for types conversion

### LINKED ISSUE #46090: SQL: incorrect DATE format when converting and GROUPing BY
ES/SQL will return a DATETIME value (instead of a DATE) when `GROUP BY` is applied:
`SELECT {fn CONVERT("@timestamp", SQL_DATE)} FROM kibana_sample_data_logs GROUP BY 1`
will return:
```
"columns" : [
    {
      "name" : """CONVERT("@timestamp", SQL_DATE)""",
      "type" : "date"
    }
  ],
  "rows" : [
    [
      "2018-07-22T00:39:02.912Z"
    ],
...
```

It works as expected without the `GROUP BY` clause (that is `"2018-07-22T00:00:00Z"`)

### PR DESCRIPTION
**Pull request by astefan**
Prior to this small change the Painless scripting was using an incorrect method for CASTing/CONVERTing a value to a certain data type. The previous approach was always returning a DATETIME data type for a DATE target type. The current change is returning the correct `DATE` type.

Fixes https://github.com/elastic/elasticsearch/issues/46090.

### CODE REVIEW

File: x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java [hunk 1]
```diff
@@ -1157,6 +1157,40 @@ public void testGetDateType() throws Exception {
...
+    public void testGetDateTypeFromAggregation() throws Exception {
+        createIndex("test");
+        updateMapping("test", builder -> builder.startObject("test_date").field("type", "date").endObject());
+
+        // 1984-05-02 14:59:12 UTC
+        Long timeInMillis = 452357952000L;
+        index("test", "1", builder -> builder.field("test_date", timeInMillis));
+
+        String timeZoneId1 = "UTC";
+        Calendar c = Calendar.getInstance(TimeZone.getTimeZone(timeZoneId1), Locale.ROOT);
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java]: I would try to use ZonedDateTime instead to reset to start of day, please check `DateUtils.asDateOnly(long millis)`.

**Review by matriv — APPROVED** [review]
LGTM, left a comment.

**Review by costin — APPROVED** [review]
LGTM.
P.S. Please add the ES version to the PR

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java [hunk 2]
```diff
@@ -548,6 +548,6 @@ public static Double stZ(Object g) {
     public static Object cast(Object value, String typeName) {
         // we call asDateTime here to make sure we handle JodaCompatibleZonedDateTime properly,
         // since casting works for ZonedDateTime objects only
-        return DataTypeConversion.convert(asDateTime(value, true), DataType.fromTypeName(typeName));
+        return DataTypeConversion.convert(asDateTime(value, true), DataType.fromSqlOrEsType(typeName));
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/whitelist/InternalSqlScriptUtils.java]: I wonder if the `DateType.fromTypeName` method shouldn't be investigated as I see it's still being used inside the code in various places related to scripting.

**Comment by astefan** [comment]
@elasticmachine run elasticsearch-ci/2

**Pull request merged**